flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/FunctionDefinition.java

```diff
@@ -40,6 +40,13 @@
 	 */
 	FunctionKind getKind();
 
+	/**
+	 * Returns the language of function this definition describes.
+	 */
+	default FunctionLanguage getLanguage() {
```

Comment:
@hequn8128 I just had a brief look at this change. I already mentioned my concern on the ML discussion. Why should users be able to override this property. `UserDefinedFunction` should mark this method `final` similar to `ScalarFunction#getKind`. What do you think?

Reply:
@twalthr Thanks a lot for your comments. Appreciated! Just share my thoughts as following:
1. `Python ScalarFunction` also extends `ScalarFunction` and it will override `getLanguage`. If it's declared as final, there is no way to override it any more. This can be seen from the implementation of [PythonFunctionCodeGenerator](https://github.com/apache/flink/blob/fc4a1df1ad48224ae846fbbdeeb21b4b7b164e02/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/codegen/PythonFunctionCodeGenerator.scala#L74).
2. Actually users could write a user-defined Python `ScalarFunction` with the `ScalarFunction` interface if they want(In this case, they need to override the getLanguage interface). [PythonFunctionCodeGenerator](https://github.com/apache/flink/blob/fc4a1df1ad48224ae846fbbdeeb21b4b7b164e02/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/codegen/PythonFunctionCodeGenerator.scala#L74) could be seen as an example how to do that, e.g. they need just implement the `PythonFunction` interface and wrap the serialized Python function inside it.

Does it make sense to you? Looking forward to your feedback.

Reply:
From an architectural point of view, I'm unsure if a Python function should be a `UserDefinedFunction`. Because it is more a system bridging function for executing Python UDFs.

As far as I can see you already do `implements PythonFunction`; isn't that enough information? I still don't understand why users need to be able to override a `getLanguage()` method. When is this method read?

Reply:
+1. If we add the `PythonFunciton` interface for Python, I think `getLanguage()` can be removed.

Reply:
`getLanguage` is used in places where it's needed to determine whether a UDF is Python not not, e.g. in [PythonScalarFunctionSplitRule](https://github.com/apache/flink/blob/2b608e3169ee416c922a8337c2ba5fa873516fde/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/rules/logical/PythonScalarFunctionSplitRule.scala#L149). As the interface `PythonFunction` has provided enough information for that purpose, so I'm +1 remove it. Have created ticket https://issues.apache.org/jira/browse/FLINK-14342 to track it.